### PR TITLE
Storing client data in tdb

### PR DESCRIPTION
**Pull request by avbelov23**
Requires #1115

Changes:

- Transfer `tfw_current_timestamp()` to library
- Keep client data in TDB (`client_db` and `client_tbl_size` options) after the last client connection is closed and use their some time (`client_lifetime` option)

### CODE REVIEW

File: tempesta_fw/client.c [hunk 1]
```diff
@@ -90,58 +69,72 @@ EXPORT_SYMBOL(tfw_client_put);
...
 
-	if (!(cli = kmem_cache_alloc(cli_cache, GFP_ATOMIC | __GFP_ZERO))) {
-		spin_unlock(&hb->lock);
+	if (unlikely(!ss_active())) {
+		TFW_DBG("reject allocation of new client after shutdown\n");
 		return NULL;
 	}
 
+	len = sizeof(*cli);
+	rec = tdb_entry_alloc(ip_client_db, key, &len);
```
[the remark this answers is not in the document]
Comment (avbelov23) [reply tempesta_fw/client.c]: On previous review was locks

File: etc/tempesta_fw.conf [hunk 2]
```diff
@@ -1115,3 +1115,38 @@
...
+# the same address is considered the same. Zero value means unlimited life time.
+#
+# Syntax:
+#   client_lifetime NUM;
+#
+# Example:
+#   client_lifetime 3600;
+#
+# Default:
+#   client_lifetime 0;
```
[the remark this answers is not in the document]
Comment (avbelov23) [reply etc/tempesta_fw.conf]: Need to choose some reasonable value

File: tempesta_fw/client.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (ikoveshnikov) [reply tempesta_fw/client.c]: That was old locks used, when `TfwClient`s were stored in hash table. That locks (hashtable of  spinlocks) wasn't synchronised with TDB and it looked like a workaround standing aside. Every time a new entry was created TDB, spinlock from the hashtable was picked and locked, then TDB search was performed, which also use some locking. 

I would like to see more elegant solution. Unlike to `cache.c`, in `client.c` we still need to insert a new entry if TDB lookup failed. Having TDB interface capable to do this in one action seems to be cleaner solution.

@krizhanovsky had more ideas on future TDB evolution, let's involve him in this discussion.

File: tempesta_fw/client.c [hunk 3]
```diff
@@ -90,58 +69,72 @@ EXPORT_SYMBOL(tfw_client_put);
...
+
+	cli = &ent->cli;
+
 	tfw_peer_init((TfwPeer *)cli, &addr);
-	hlist_add_head(&cli->hentry, &hb->list);
-	cli->hb_lock = &hb->lock;
 	if (init)
 		init(cli);
 
-	atomic64_inc(&act_cli_n);
```
Comment (ikoveshnikov) [inline tempesta_fw/client.c]: The variable was required for `tfw_cli_wait_release()`.

**Comment by avbelov23** [comment]
DONE

File: tempesta_fw/client.c [hunk 4]
```diff
@@ -181,45 +197,104 @@ tfw_cli_wait_release(void)
...
+tfw_client_stop(void)
+{
+	if (tfw_runstate_is_reconfig())
+		return;
+	if (ip_client_db)
+		tdb_close(ip_client_db);
+}
+
+static int
+tfw_client_lifetime(TfwCfgSpec *cs, TfwCfgEntry *ce)
```
Comment (krizhanovsky) [inline tempesta_fw/client.c]: We use `tfw_cfgop_` prefixes for configuration handlers

File: tempesta_fw/client.c [hunk 5]
```diff
@@ -113,28 +112,53 @@ tfw_client_obtain(struct sock *sk, void (*init)(TfwClient *))
...
 
-	atomic64_inc(&act_cli_n);
-	TFW_INC_STAT_BH(clnt.online);
 	TFW_DBG("new client: cli=%p\n", cli);
 	TFW_DBG_ADDR("client address", &cli->addr, TFW_NO_PORT);
 
 found:
+	if (!atomic_read(&cli->conn_users)) {
+		atomic64_inc(&act_cli_n);
+		TFW_INC_STAT_BH(clnt.online);
```
[the remark this answers is not in the document]
Comment (avbelov23) [reply tempesta_fw/client.c]: The client now appears online when it is allocated or got from tdb and conn_users is zero

File: tempesta_fw/client.c [hunk 6]
```diff
@@ -152,22 +145,14 @@ EXPORT_SYMBOL(tfw_client_obtain);
 int
 tfw_client_for_each(int (*fn)(TfwClient *))
 {
-	int i, r = 0;
-
-	for (i = 0; i < CLI_HASH_SZ && !r; ++i) {
-		TfwClient *c;
-		CliHashBucket *hb = &cli_hash[i];
+	int r = 0;
+	/*TfwClient *c;
```
[the remark this answers is not in the document]
Comment (avbelov23) [reply tempesta_fw/client.c]: This code is not associated with deleting clients. It is associated with the closing of all connections of all clients. Without going through all the records and closing of all connections will crash. Need `tdb_for_each_rec`

File: tempesta_fw/client.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (aleksostapenko) [reply tempesta_fw/client.c]: Yeah, missed that.

**Pull request closed without merging**